resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/builders/AutoBuildJobRescheduleTest.java

```diff
@@ -0,0 +1,200 @@
+
+	/**
+	 * The barrier to synchronize the builder and the concurrent workspace
+	 * modification
+	 */
+	private TestBarrier2 barrier = new TestBarrier2();
+
+	/**
+	 * A file in the workspace that we modify to trigger builds
+	 */
+	private IFile file;
+
+	/**
+	 * A thread-safe flag to record the rescheduling of the AutoBuildJob
+	 */
+	private AtomicBoolean rescheduled = new AtomicBoolean(false);
+
+	/**
+	 * A somewhat illegal {@link ISchedulingRule} which satisfies the sanity checks
+	 * in {@code JobManager.validateRule()} but still does not prevent that
+	 * {@link AutoBuildJobRescheduleTest#file} is modified while the
+	 * {@code AutoBuildJob} is running.
+	 *
+	 * @since 3.2
+	 */
+	public class SpoiledSchedulingRule implements ISchedulingRule {
+		@Override
+		public boolean isConflicting(ISchedulingRule rule) {
+			return rule == this || rule == file;
+		}
+
+		@Override
+		public boolean contains(ISchedulingRule rule) {
+			return rule == this || rule == file;
+		}
+	}
+
+	/**
+	 * An {@link IJobChangeListener} that reacts to the AutoBuildJob being scheduled
+	 * or done.
+	 *
+	 * Since the AutoBuildJob is not visible, we check against the class name.
+	 */
+	private IJobChangeListener jobChangeListener = new JobChangeAdapter() {
+		@Override
+		public void scheduled(IJobChangeEvent event) {
+			if ("AutoBuildJob".equals(event.getJob().getClass().getSimpleName())) {
+				rescheduled.set(true);
+			}
+		}
+
+		@Override
+		public void done(IJobChangeEvent event) {
+			if ("AutoBuildJob".equals(event.getJob().getClass().getSimpleName())) {
+				barrier.setStatus(AUTOBUILD_DONE);
+			}
+		}
+	};
+
+	public AutoBuildJobRescheduleTest(String testName) {
+		super(testName);
+	}
+
+	/**
+	 * Setup the test by creating a project and a file. We also add a special
+	 * builder that suspends execution using the {@link #barrier}.
+	 */
+	@Override
+	protected void setUp() throws Exception {
+		super.setUp();
+
+		var project = getWorkspace().getRoot().getProject("PROJECT");
+		file = project.getFile("file");
+		try {
+			// Turn auto-building off to prevent too early builds
+			setAutoBuilding(false);
+			// Create and open the project
+			project.create(getMonitor());
+			project.open(getMonitor());
+			// Create the file
+			ensureExistsInWorkspace(file, getRandomContents());
+
+			// Add the EmptyDeltaBuilder. The behavior of that builder is configured below.
+			IProjectDescription desc = project.getDescription();
+			desc.setBuildSpec(new ICommand[] { createCommand(desc, EmptyDeltaBuilder.BUILDER_NAME, "ProjectBuild") });
+			project.setDescription(desc, getMonitor());
+			project.build(IncrementalProjectBuilder.INCREMENTAL_BUILD, getMonitor());
+
+			// Perform the first build. This instantiates the EmptyDeltaBuilder and makes
+			// its instance accessible.
+			setAutoBuilding(true);
+
+			// Configure the behavior of the builder
+			EmptyDeltaBuilder.getInstance().setRuleCallback(new BuilderRuleCallback() {
```

Comment:
shouldn't that happen before autobuild?

Reply:
Actually doesn’t matter as we had an explicit incremental build before (to let the system instantiate EmptyDataBuilder and make its instance available). Therefore, setting the autobuild to true here will not invoke the builder just yet.